wireguard-go-rs/src/lib.rs

```diff
@@ -1,70 +1,265 @@
 #![cfg(unix)]
+
+use core::slice;
+use std::{
+    ffi::{c_char, c_void, CStr},
+    mem::ManuallyDrop,
+};
+use zeroize::Zeroize;
+
 pub type Fd = std::os::unix::io::RawFd;
-use std::ffi::{c_char, c_void};
+
 pub type WgLogLevel = u32;
+
 pub type LoggingCallback =
     unsafe extern "system" fn(level: WgLogLevel, msg: *const c_char, context: *mut c_void);
 
-extern "C" {
-    /// Creates a new wireguard tunnel, uses the specific interface name, MTU and file descriptors
-    /// for the tunnel device and logging.
-    ///
-    /// Positive return values are tunnel handles for this specific wireguard tunnel instance.
-    /// Negative return values signify errors. All error codes are opaque.
-    #[cfg(not(target_os = "android"))]
-    pub fn wgTurnOn(
-        mtu: isize,
-        settings: *const i8,
-        fd: Fd,
-        logging_callback: Option<LoggingCallback>,
-        logging_context: *mut c_void,
-    ) -> i32;
+/// A wireguard-go tunnel
+pub struct Tunnel {
+    /// wireguard-go handle to the tunnel.
+    handle: i32,
+}
 
-    // Android
-    #[cfg(target_os = "android")]
-    pub fn wgTurnOn(
-        settings: *const i8,
-        fd: Fd,
+// NOTE: Must be kept in sync with libwg.go
+// NOTE: must be kept in sync with `result_from_code`
+// INVARIANT: Will aways be represented as a negative i32
+#[repr(i32)]
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Something went wrong.")]
+    GeneralFailure = -1,
+
+    #[error("Something went wrong, but trying again might help.")]
+    IntermittentFailure = -2,
+
+    #[error("An argument you provided was invalid.")]
+    InvalidArgument = -3,
+
+    #[error("The tunnel handle did not refer to an existing tunnel.")]
+    UnknownTunnel = -4,
+
+    #[error("The provided public key did not refer to an existing peer.")]
+    UnknownPeer = -5,
+
+    #[error("Something went wrong when enabling DAITA.")]
+    EnableDaita = -6,
+
+    #[error("`libwg` provided an unknown error code. This is a bug.")]
+    Other = i32::MIN,
+}
+
+impl Tunnel {
+    // TODO: this function is supposed to be a safe wrapper, but as clippy points out, 
+    // the logging_context is a *mut, which may unsafely be dereferenced by the callback.
```

Comment:
Why does the function have to take logging_context as a `*mut c_void` and not `logging_context: &mut LoggingContext`?

Reply:
can't pass rust references through ffi,e
i'm thinking we can drop the *mut though and just pass a u64 instead :thinking:

Reply:
But can't the `turn_on` fn handle the conversion from `&mut LoggingContext` to `*mut c_void`?

Reply:
If we're passing a `&mut` over ffi (even as a *mut), we would need to make sure the go-code doesn't violate any of the rust rules about references, which might be tricky.

Thing is though, right now we're not actually passing a pointer as the logging context, instead we're just passing a `u32` by casting it to a  `*mut c_void`. For our use-case, there's no need to use pointers.

Reply:
Nice!